### PR TITLE
feat: "initial commit"

### PR DESCRIPTION
**Pull request by LacticWhale**
Initial commit. Code examples are included.

Known issues: None.

### CODE REVIEW

File: lib/src/state.dart [hunk 1]
```diff
@@ -0,0 +1,10 @@
+enum State {
+  none,
+  closed,
+  handshake,
```
Comment (Zekfad) [inline lib/src/state.dart:4]: use ing

File: lib/src/state.dart [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+  authentication,
```
Comment (Zekfad) [inline lib/src/state.dart:5]: use ing

File: lib/src/state.dart [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+  ready,
+  connected,
+  // binded, // To hard to understand
```
Comment (Zekfad) [inline lib/src/state.dart:8]: because it's "bound"

Comment (Zekfad) [reply lib/src/state.dart:8]: and maybe better to use "listening"?

File: lib/src/tcp_connection.dart [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+  late final InternetAddress clientAddress;
+  late final int clientPort;
+
+  TcpConnection.extend(this._connection) {
+    targerAddress = _connection.desiredAddress!;
+    targerPort = _connection.desiredPort!;
+
+    clientAddress = _connection.socket.remoteAddress;
+    clientPort = _connection.socket.remotePort;
```
Comment (Zekfad) [inline lib/src/tcp_connection.dart:31]: use initialization

File: lib/src/tcp_connection.dart [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  StreamSink<Uint8List> get outcomingSink => _outcomingStreamController.sink;
+
+  late final Socket targetSocket;
+  late final InternetAddress targerAddress;
+  late final int targerPort;
+
+  late final InternetAddress clientAddress;
+  late final int clientPort;
+
+  TcpConnection.extend(this._connection) {
```
Comment (Zekfad) [inline lib/src/tcp_connection.dart:26]: better to say "from"

File: lib/src/tcp_connection.dart [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      _connection.socket.add(data);
+    });
+  }
+
+  @override
+  Future<void> reject(SocksResponse message) async => await _connection.reject(
+          message: Uint8List.fromList([
+        0x05,
+        message.index,
+      ]));
```
Comment (Zekfad) [inline lib/src/tcp_connection.dart:43]: format code

File: lib/src/tcp_connection.dart [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        _connection.reject(message: Uint8List.fromList([0x05, 0x05]));
+        rethrow;
+      }
+    }
+
+    _connection.socket.add([
+      0x05,
+      0x00,
+      0x00,
+      _connection.getAddressTypeAsBite(targerAddress.type),
```
Comment (Zekfad) [inline lib/src/tcp_connection.dart:59]: bite?

Comment (Zekfad) [reply lib/src/tcp_connection.dart:59]: and just name it getAddressType, and isn't it must be a private method?

File: lib/src/tcp_connection.dart [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  @override
+  Future<void> forward() async {
+    await accept(connect: true);
+
+    final target = await Socket.connect(targerAddress, targerPort);
+
+    target.listen((Uint8List data) {
+      outcomingSink.add(data);
```
Comment (Zekfad) [inline lib/src/tcp_connection.dart:79]: name it clientSink

File: lib/src/tcp_connection.dart [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    _connection.stream.listen(_incomingStreamController.add);
+
+    return;
+  }
+
+  @override
+  Future<void> forward() async {
+    await accept(connect: true);
+
+    final target = await Socket.connect(targerAddress, targerPort);
```
Comment (Zekfad) [inline lib/src/tcp_connection.dart:76]: typo targer

File: lib/src/typed_connections.dart [hunk 10]
```diff
@@ -0,0 +1,9 @@
+export 'tcp_connection.dart';
+export 'udp_connection.dart';
+
+import 'errors.dart';
+
+abstract class Connection {
+  Future<void> reject(SocksResponse message);
+  Future<void> forward();
+}
```
Comment (Zekfad) [inline lib/src/typed_connections.dart:9]: accept must have the same signature for both (with `{bool connect}`)

File: lib/src/udp_connection.dart [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+  Stream<UDPPacket> get clientPackets => _incomingClientStreamController.stream;
+
+  final _incomingTargetStreamController = StreamController<Datagram>();
+  Stream<Datagram> get remotePackets => _incomingTargetStreamController.stream;
+
+  UdpConnection.extend(this._connection) {
+    if (_connection.state != State.associated) {
+      throw ArgumentError.value(
+          _connection, 'Connection doesn\'t have associated upd socket');
```
Comment (Zekfad) [inline lib/src/udp_connection.dart:31]: UDP must be capitalized

File: lib/src/state.dart [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+  associated,
+}
```
Comment (Zekfad) [inline lib/src/state.dart:10]: add docs

File: lib/src/udp_connection.dart [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+  int sendToTarget(Uint8List data, InternetAddress remoteAddress, int port) =>
+      _targetSocket.send(data, remoteAddress, port);
+
+  @override
+  Future<void> reject(SocksResponse message) async => await _connection.reject(
+          message: Uint8List.fromList([
+        0x05,
+        message.index,
+      ]));
```
Comment (Zekfad) [inline lib/src/udp_connection.dart:49]: format

File: lib/src/udp_connection.dart [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+          SocksResponse.connectionRefused.index,
+        ]),
+      );
+      rethrow;
+    }
+    if (connect) {
+      _targetSocket.listen((RawSocketEvent event) {
+        if (event != RawSocketEvent.read) return;
+
+        Datagram? dg = _targetSocket.receive();
```
Comment (Zekfad) [inline lib/src/udp_connection.dart:72]: bad name

File: lib/src/udp_connection.dart [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+import 'udp_packet.dart';
+import 'typed_connections.dart';
+
+class UdpConnection extends Connection {
+  final ConnectionBased _connection;
+
+  late final InternetAddress clientAddress;
+  late final int clientPort;
+
+  final _incomingClientStreamController = StreamController<UDPPacket>();
```
Comment (Zekfad) [inline lib/src/udp_connection.dart:18]: name it `_clientPackets`

File: lib/src/udp_packet.dart [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    int offset = 2;
+
+    frags = bytes.getUint8(offset++);
+
+    _type = bytes.getUint8(offset++);
+    remoteAddressType = {
+      1: InternetAddressType.IPv4,
+      3: InternetAddressType.unix,
+      4: InternetAddressType.IPv6,
+    }[_type]!;
```
Comment (Zekfad) [inline lib/src/udp_packet.dart:30]: this map is const

File: pubspec.yaml [hunk 17]
```diff
@@ -0,0 +1,13 @@
+name: socks5
+description: A starting point for Dart libraries or applications.
```
Comment (Zekfad) [inline pubspec.yaml:2]: add description

**Review by Zekfad — CHANGES_REQUESTED** [review]
some changes required

File: examples/1_simple_proxy.dart [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+
+main() {
+  // create server object
+  var proxy = Socks5Server();
+
+  // listen to all tcp and udp connections
+  proxy.connections.listen((Connection connection) async {
+    try {
+      // forward everythingtry
+      await connection.forward();
```
[the remark this answers is not in the document]
Comment (LacticWhale) [reply examples/1_simple_proxy.dart:13]: yeah

File: lib/src/connection.dart [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+  AuthHandler? _authHandler;
+  set authHandler(AuthHandler handler) => _authHandler = handler;
+
+  late final Stream<Uint8List> stream;
+  late final ChunkedStreamReader<int> _data;
+
+  ConnectionBased(this.socket) {
+    stream = socket.asBroadcastStream();
+    _data = ChunkedStreamReader(stream);
+  }
```
[the remark this answers is not in the document]
Comment (LacticWhale) [reply lib/src/connection.dart:33]: 👍

File: lib/src/connection.dart [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+    String username = utf8.decode((await _data.readChunk(ul)));
+    int pl = await _readByte();
+    String password = utf8.decode((await _data.readChunk(pl)));
+
+    if (_authHandler!(username, password)) {
+      state = State.ready;
+      socket.add([1, 0]);
+      await socket.flush();
+    } else {
+      reject(message: Uint8List.fromList([1, 1]));
```
[the remark this answers is not in the document]
Comment (LacticWhale) [reply lib/src/connection.dart:81]: no

**Review by Zekfad — CHANGES_REQUESTED** [review]

File: lib/src/server.dart [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+    if (a.codeUnits.length != b.codeUnits.length) {
+      return false;
+    }
+
+    var r = 0;
+    for (int i = 0; i < a.codeUnits.length; i++) {
+      r |= a.codeUnitAt(i) ^ b.codeUnitAt(i);
+    }
+    return r == 0;
+  }
```
Comment (Zekfad) [inline lib/src/server.dart:58]: extract to file

File: lib/src/tcp/tcp_connection.dart [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+import '../typed_connections.dart';
+
+class TcpConnection extends Connection {
+  final SocksConnection _connection;
+  Socket? target;
+
+  final _incomingStreamController = StreamController<Uint8List>();
+  Stream<Uint8List> get incomingStream => _incomingStreamController.stream;
+
+  final _outcomingStreamController = StreamController<Uint8List>();
```
Comment (Zekfad) [inline lib/src/tcp/tcp_connection.dart:21]: outgoing

File: lib/src/tcp/tcp_connection.dart [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+    _connection.socket.add([
+      _connection.socksVersion,
+      0x00,
+      0x00,
+      targetAddress.type.byte,
+      ...targetAddress.rawAddress,
+      ...[
+        (targetPort & 0x00ff) >> 8,
+        targetPort & 0xff,
+      ]
```
Comment (Zekfad) [inline lib/src/tcp/tcp_connection.dart:69]: why?

File: lib/src/udp/udp_connection.dart [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+  UdpConnection.from(this._connection)
+      : assert(
+          _connection.associatedAddress != null,
+          "Connection hasn't yet been established",
+        ),
+        clientAddress = _connection.associatedAddress!,
+        assert(
+          _connection.associatedPort != null,
+          "Connection hasn't yet been established",
+        ),
```
Comment (Zekfad) [inline lib/src/udp/udp_connection.dart:27]: just use `_connection.associatedAddress != null && _connection.associatedPort != null`

File: lib/src/udp/udp_connection.dart [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    _connection.socket.add([
+      _connection.socksVersion,
+      0x00, // reserved
+      0x00, // reserved
+      _connection.socket.address.type.byte,
+      // tell client to connect to same ip as tcp socket
+      ..._connection.socket.address.rawAddress,
+      ...[(port! & 0xff00) >> 8, port! & 0x00ff],
```
Comment (Zekfad) [inline lib/src/udp/udp_connection.dart:46]: why?

File: lib/src/udp/udp_connection.dart [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+          "Connection hasn't yet been established",
+        ),
+        clientPort = _connection.associatedPort!,
+        super(_connection.socket.address);
+
+  @override
+  Future<void> accept([bool connect = false]) async {
+    await super.ready();
+
+    if (port == null) {
```
Comment (Zekfad) [inline lib/src/udp/udp_connection.dart:35]: what is that port?

File: lib/src/udp/udp_connection.dart [hunk 27]
```diff
@@ -0,0 +1,86 @@
...
+      _connection.socket.address.type.byte,
+      // tell client to connect to same ip as tcp socket
+      ..._connection.socket.address.rawAddress,
+      ...[(port! & 0xff00) >> 8, port! & 0x00ff],
+    ]);
+  }
+
+  @override
+  Future<void> forward() async {
+    accept(true);
```
Comment (Zekfad) [inline lib/src/udp/udp_connection.dart:52]: no await?
also what if bind failed?

File: lib/src/internet_address_type_extention.dart [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+      return 1;
+    } else if (this == InternetAddressType.unix) {
+      return 3;
+    } else {
+      return 4;
+    }
+  }
+}
+
+InternetAddressType internetAddressTypeFrom(int byte) {
```
[the remark this answers is not in the document]
Comment (Zekfad) [reply lib/src/internet_address_type_extention.dart:15]: shouldn't it be in SocksUdpPacket?

File: lib/src/udp/udp_connection.dart [hunk 29]
```diff
@@ -0,0 +1,86 @@
...
+        clientAddress,
+        clientPort,
+      );
+    });
+
+    stream.listen((Datagram datagram) {
+      if (datagram.address != _connection.associatedAddress ||
+          datagram.port != _connection.associatedPort) return;
+
+      final packet = SocksUdpPacket.tryParse(datagram);
```
Comment (Zekfad) [inline lib/src/udp/udp_connection.dart:73]: tryParse means nullable, parse is tryParse call with throw state error on null value

File: lib/src/udp/udp_server.dart [hunk 30]
```diff
@@ -0,0 +1,42 @@
+import 'dart:async';
+import 'dart:io';
+import 'dart:typed_data';
+
+class UdpServer {
+  final Future<RawDatagramSocket> _socketFuture;
```
Comment (Zekfad) [inline lib/src/udp/udp_server.dart:6]: bad name

File: lib/src/udp/udp_server.dart [hunk 31]
```diff
@@ -0,0 +1,42 @@
...
+
+  final _remoteStream = StreamController<Datagram>();
+  Stream<Datagram> get stream => _remoteStream.stream;
+
+  UdpServer(InternetAddress address)
+      : _socketFuture = RawDatagramSocket.bind(address, 0) {
+    _socketFuture.then((socket) {
+      _socket = socket;
+      _socket!.listen((RawSocketEvent event) {
+        print(event);
```
Comment (Zekfad) [inline lib/src/udp/udp_server.dart:19]: ?

File: lib/src/udp/udp_server.dart [hunk 32]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  Future<void> ready() async {
+    await _socketFuture;
+    return;
+  }
+
+  Future<int> send(Uint8List data, InternetAddress address, int port) async {
+    if (_socket == null) {
+      throw SocketException('Server haven\'t yet been opened.');
```
Comment (Zekfad) [inline lib/src/udp/udp_server.dart:37]: bad wording

**Review by Zekfad — CHANGES_REQUESTED** [review]
also add comments with short explanation of example before main function

File: examples/3_tcp_filtering_proxy.dart [hunk 33]
```diff
@@ -0,0 +1,64 @@
+import 'dart:convert';
+import 'dart:io';
+import 'dart:typed_data';
+
+import 'package:socks5/socks5.dart';
+
+bool testIPMask(InternetAddress address, List<int> mask) {
+  if (mask.length < 4) throw RangeError('Mask size must be 4');
```
[the remark this answers is not in the document]
Comment (LacticWhale) [reply examples/3_tcp_filtering_proxy.dart:8]: i forgor

File: examples/4_tcp_spoofing.dart [hunk 34]
```diff
@@ -0,0 +1,48 @@
...
+
+void main() {
+  // create server object
+  var proxy = Socks5Server();
+
+  // listen to all tcp and udp connections
+  proxy.connections.listen((Connection connection) async {
+    if (connection is TcpConnection) {
+      // forward everithing except connections to target 93.184.216.34
+      if (connection.targetAddress != InternetAddress('93.184.216.34')) {
```
[the remark this answers is not in the document]
Comment (LacticWhale) [reply examples/4_tcp_spoofing.dart:14]: example.com, cause unix addresses works only on linux, mac

File: examples/5_udp_spoofing.dart [hunk 35]
```diff
@@ -0,0 +1,42 @@
...
+  proxy.connections.listen((Connection connection) async {
+    if (connection is TcpConnection) {
+      // we don't want tcp proxy, reject everything
+      return connection.reject(CommandResponse.unsupported);
+    } else if (connection is UdpConnection) {
+      // accept connection
+      await connection.accept();
+
+      connection.stream.listen((Datagram datagram) {
+        // reject datagrams from unknown clients
```
[the remark this answers is not in the document]
Comment (LacticWhale) [reply examples/5_udp_spoofing.dart:20]: 👍

File: examples/6.1_udp_reading.dart [hunk 36]
```diff
@@ -0,0 +1,65 @@
...
+        print(datagram.data.map((e) => e.toRadixString(16)));
+
+        // send datagram with socks5 header to client
+        connection.send(
+          // append socks5 header to data in datagram
+          Socks5Server.createUdpFrame(
+            datagram.data,
+            datagram.address,
+            datagram.port,
+          ),
```
[the remark this answers is not in the document]
Comment (LacticWhale) [reply examples/6.1_udp_reading.dart:40]: datagram.data = socks5header aka frame + data

File: lib/src/enums/authentication_method.dart [hunk 37]
```diff
@@ -0,0 +1,21 @@
+import 'message.dart';
+
+enum AuthenticationMethod implements Message {
+  /// 0x00 NO AUTHENTICATION REQUIRED
+  noAuthentication(0x00),
+
+  // /// 0x01 GSSAPI
```
[the remark this answers is not in the document]
Comment (LacticWhale) [reply lib/src/enums/authentication_method.dart:7]: I messed up with values

File: lib/src/tcp/tcp_connection.dart (hunk 23 above)
Comment (LacticWhale) [reply lib/src/tcp/tcp_connection.dart:69]: ?

File: lib/src/udp/udp_connection.dart (hunk 25 above)
Comment (LacticWhale) [reply lib/src/udp/udp_connection.dart:46]: ?

File: lib/src/udp/udp_connection.dart (hunk 29 above)
Comment (LacticWhale) [reply lib/src/udp/udp_connection.dart:73]: noted

File: lib/src/udp/udp_server.dart (hunk 31 above)
Comment (LacticWhale) [reply lib/src/udp/udp_server.dart:19]: this way _socket isn't future

File: examples/4_tcp_spoofing.dart (hunk 34 above)
[the remark this answers is not in the document]
Comment (Zekfad) [reply examples/4_tcp_spoofing.dart:14]: unix address is actually a file fyi...

File: lib/src/udp/udp_server.dart (hunk 31 above)
Comment (Zekfad) [reply lib/src/udp/udp_server.dart:19]: that's bad naming over there then

**Pull request closed without merging**